File: src/main/java/com/fic/udc/es/padel/services/UserService.java
package com.fic.udc.es.padel.services;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Set;

import com.fic.udc.es.padel.model.entities.Game;
import com.fic.udc.es.padel.model.entities.Schedule;
import com.fic.udc.es.padel.model.entities.User;
import com.fic.udc.es.padel.model.exceptions.DuplicateInstanceException;
import com.fic.udc.es.padel.model.exceptions.IncorrectLoginException;
import com.fic.udc.es.padel.model.exceptions.IncorrectPasswordException;
import com.fic.udc.es.padel.model.exceptions.InstanceNotFoundException;

public interface UserService {
	
	void signUp(User user) throws DuplicateInstanceException;
	
	User login(String userName, String password) 
			throws IncorrectLoginException;
	
	User loginFromId(Long id) throws InstanceNotFoundException;
	
	User updateProfile(Long id, String firstName, String lastName1, 
			String lastName2, boolean state, String position, float level, 
			Set<Schedule> schedules) throws InstanceNotFoundException;
	
	void changePassword(Long id, String oldPassword, String newPassword)
			throws InstanceNotFoundException, IncorrectPasswordException;
	
	void changeLevel(Long id, float newLevel) throws InstanceNotFoundException;
	
	User getUserById(Long id) throws InstanceNotFoundException;
	
	Block<User> getAllUsers(int page, int size);
	
	List<User> findUserByLevelAndDate(float minLevel, float maxLevel, LocalDateTime date);

}

File: frontend/src/backend/userService.js
import {config, appFetch, setServiceToken, getServiceToken, removeServiceToken, setReauthenticationCallback} from './appFetch';

export const login = (userName, password, onSuccess, onErrors, reauthenticationCallback) =>
    appFetch('/users/login', config('POST', {userName, password}),
        authenticatedUser => {
            setServiceToken(authenticatedUser.serviceToken);
            setReauthenticationCallback(reauthenticationCallback);
            onSuccess(authenticatedUser);
        }, 
        onErrors);

export const tryLoginFromServiceToken = (onSuccess, reauthenticationCallback) => {

    const serviceToken = getServiceToken();

    if (!serviceToken) {
        onSuccess();
        return;
    }

    setReauthenticationCallback(reauthenticationCallback);

    appFetch('/users/loginFromServiceToken', config('POST'),
        authenticatedUser => onSuccess(authenticatedUser),
        () => removeServiceToken()
    );

}

export const signUp = (user, onSuccess, onErrors, reauthenticationCallback) => {
    appFetch('/users/signUp', config('POST', user), 
        authenticatedUser => {
            setServiceToken(authenticatedUser.serviceToken);
            setReauthenticationCallback(reauthenticationCallback);
            onSuccess(authenticatedUser);
        }, 
        onErrors);

}

export const logout = () => removeServiceToken();

export const updateProfile = (user, onSuccess, onErrors) =>
    appFetch(`/users/${user.id}`, config('PUT', user),
        onSuccess, onErrors);

export const changePassword = (id, oldPassword, newPassword, onSuccess,
    onErrors) =>
    appFetch(`/users/${id}/changePassword`, 
        config('POST', {oldPassword, newPassword}),
        onSuccess, onErrors);

export const getAllUsers = ({page, login, firstName, minLevel, maxLevel}, onSuccess) => {
    var params = `?page=${page}`;
    if(login){
        params += `&login=${login}`
    }
    if(firstName){
        params += `&name=${firstName}`
    }
    if(minLevel){
        params += `&minLevel=${minLevel}`
    }
    if(maxLevel){
        params += `&maxLevel=${maxLevel}`
    }
    appFetch(`/users` + params, config('GET'), onSuccess);
}

export const findUserById = (id, onSuccess) => {
    appFetch(`/users/${id}`, config('GET'), onSuccess);
}

export const changeLevel = (id, level, onSuccess, onErrors) => {

    appFetch(`/users/${id}/changeLevel`, config('POST', level), 
        onSuccess, onErrors);
}

export const getAllSchedules = (userId, onSuccess) => {
    appFetch(`/users/${userId}/schedules`, config('GET'), onSuccess);
}

export const addSchedule = (userId, schedule, onSuccess) => {
    appFetch(`/users/${userId}/schedules`, config('POST', schedule), onSuccess);
}

export const findUsersByLevelAndDate = (minimunLevel, maximunLevel, date, onSuccess) => {
    appFetch(`/users/filtered?minLevel=${minimunLevel}&maxLevel=${maximunLevel}&millis=${date}`, config('GET'), onSuccess);
}

export const deleteByScheduleId = (scheduleId, userId, onSuccess) => {
    appFetch(`/users/${userId}/schedules/${scheduleId}`, config('DELETE'), onSuccess);
}